lang/wasm_instance_cache.hh

```diff
@@ -0,0 +1,326 @@
+        auto memory_export = instance.get(store, "memory");
+        if (!memory_export) {
+            throw wasm::exception("memory export not found - please export `memory` in the wasm module");
+        }
+        auto memory = std::get<wasmtime::Memory>(*memory_export);
+
+        return wasm::wasm_instance{.store=std::move(store), .instance=std::move(instance), .func=std::move(*func), .memory=std::move(memory)};
+    }
+
+    // lru must not be empty, and its elements must refer to entries in _cache
+    void evict_lru() noexcept {
+        auto& entry = _lru.front();
+        _total_size -= entry.instance_size;
+        _lru.pop_front();
+        auto it = _cache.find(entry.key);
+        _cache.erase(it);
+    }
+
+    void on_timer() noexcept {
+        auto now = seastar::lowres_clock::now();
+        while (!_lru.empty() && _lru.front().timestamp + _timer_period < now) {
+            evict_lru();
+        }
+    }
+
+public:
+    seastar::future<value_type> get(const key_type& key, wasm::context& ctx) {
+        cache_key_type key_pair(seastar::current_scheduling_group(), key);
+        auto& i = _cache[key_pair];
+        if (!i.instance) {
+            ++shard_stats().cache_misses;
+            i.instance = seastar::make_lw_shared<wasm_instance>(load(ctx));
+            _lru.emplace_back(key_pair, seastar::lowres_clock::now(), i.instance->memory.size(i.instance->store));
+            i.it = --_lru.end();
+            _total_size += i.it->instance_size;
+        } else {
+            ++shard_stats().cache_hits;
+        }
+        // cache entry for key_pair shouldn't be removed until we 'recycle' the instance
+        if (!i.mutex.try_lock()) {
+            // the instance is in use
+            ++shard_stats().cache_blocks;
+        } else {
+            i.mutex.unlock();
+            // because we don't want to remove an instance in use, and also because we can't 
+            // track its size efficiently, we remove it from lru and subtract its size from
+            // the total size until it is no longer used
+            _total_size -= i.it->instance_size;
+            _lru.erase(i.it);
+        }
+        return i.mutex.lock().then([this, &i, &ctx] {
+            if (!i.instance) {
+                i.instance = seastar::make_lw_shared<wasm_instance>(load(ctx));
+            }
+            return make_ready_future<value_type>(i.instance);
+        });
+    }
+
+    void reset(const key_type& key) noexcept {
+        auto key_pair = lookup_cache_key_type(seastar::current_scheduling_group(), key);
+        auto it = _cache.find(key_pair);
+        it->second.instance = nullptr;
+    }
+
+    void recycle(const key_type& key) noexcept {
+        auto it = _cache.find(lookup_cache_key_type(seastar::current_scheduling_group(), key));
+        if (it == _cache.end()) {
+            return;
+        }
+        auto& [key_pair, i] = *it;
+        i.mutex.unlock();
+        if (i.mutex.try_lock()) {
+            i.mutex.unlock();
```

Comment:
This is an odd sequence.

Reply:
> True, but it's the only method of checking if there are waiters on this mutex.

Then a comment is in order. Not only what you're doing here, but also why.

> Actually an idea just came upon me that I'm not sure is completely wrong. Would it be ok to evict only the wasm instance as a result of lru, and leave the mutexes indefinitely? With that, we could avoid all such sequences and there would be less misunderstandings of when are we tracking the size, and when is the instance in the lru

Wouldn't then we have mutexes accumulating in memory? Suppose @nyh writes a test that creates a UDF, uses it once, then deletes it, and repeats the sequence indefinitely. Would it oom?

Perhaps the mutexes should live in a separate hash, and only exist for instances in use. Then the instance can disappear independently of whether someone is waiting for it. We don't have a weak map though, so it's hard to arrange to the mutex to clean itself up.

Reply:
> > True, but it's the only method of checking if there are waiters on this mutex.
> 
> Then a comment is in order. Not only what you're doing here, but also why.

I've added a comment

> Wouldn't then we have mutexes accumulating in memory? Suppose @nyh writes a test that creates a UDF, uses it once, then deletes it, and repeats the sequence indefinitely. Would it oom?

Yes, it would, so this is probably not a good idea for now. I was thinking that maybe we could limit the number of mutexes stored simultaneously in the cache by the number of functions in the database - one (or a few for different scheduling groups) didn't seem like a big price to pay.
That would however require removing the indexes for dropped functions, so we need to add to the process of dropping a UDF a step for removing the corresponding cache instances.


... actually, this needs to be added anyway. Currently, if we create a UDF, use it, drop it, and create a new one with the same name and try to use it, we will incorrectly use the cached instance for the previous UDF.

Considering that I need to add the handling of instances for UDFs that have been dropped to prevent the use of incorrect instances, the number of mutexes will in fact be limited by the number of UDFs that have been used but not dropped multiplied by the number of scheduling groups it has been used in. I'm not sure how we're handling the scheduling groups - can we have a test where we create a huge number of them? If so we still may go oom, but if not the number of mutexes for UDF instaces doesn't look that big

Reply:
> I'm not sure how we're handling the scheduling groups - can we have a test where we create a huge number of them?

You can assume that the number of scheduling groups is bounded by a small constant, e.g. 16.